### ISSUE TITLE
new Cart(... does not always dispatch expected event  CartCreatedEvent 

### ISSUE THREAD

**Issue by speedupmate** [shopware/platform#1839]
**Description**  

When adding items to cart from frontend or calling some methods internally CartCreatedEvent on cart creation time is not dispatched 

When adding items to cart , cart is initiated in: 

https://github.com/shopware/platform/blob/730a28bb13c6c64a14e110a9c9af6fecff5bf444/src/Core/Checkout/Cart/Processor.php#L52 

also 

https://github.com/shopware/platform/blob/730a28bb13c6c64a14e110a9c9af6fecff5bf444/src/Core/Checkout/Cart/CartRuleLoader.php#L58

lacks the CartCreatedEvent . Whoever expects to subscribe to cart creation and expect that this will be fired each time cart object is created will have places that fail to trigger the event 

Cart is also created in and event triggered :
```
./Core/Checkout/Cart/SalesChannel/CartLoadRoute.php:88:        $cart = new Cart($name, $token);
./Core/Checkout/Cart/SalesChannel/CartService.php:105:        $cart = new Cart($name, $token);
```

**Possible Solution**  

dispatch the missing event in all places new Cart(.. is initiated. 

`$this->eventDispatcher->dispatch(new CartCreatedEvent($cart));`

**Comment by OliverSkroblin** [comment]
Hi,

you are right with the following place `\Shopware\Core\Checkout\Cart\CartRuleLoader::loadByToken`. Here the event is missing and must be added afterwards.

However, in `platform/src/Core/Checkout/Cart/Processor.php` we will not implement this as only a transfer object is created here. 

The `CartCreatedEvent` should reflect whenever the user creates a new cart and not when the object instance is created.

**Comment by speedupmate** [comment]
@OliverSkroblin why oh why :) cart is still created in this transfer object . This results in a fact that no CartCreatedEvent is dispatched and other cart related events like BeforeLineItemAddedEvent AfterLineItemAddedEvent etc are still dispatched. Like egg before hen :)

**Comment by speedupmate** [comment]
Also it seems like the method is bit hardcoded and just adding a event there would not solve all the issues in this method? 

```
public function loadByToken(SalesChannelContext $context, string $cartToken): RuleLoaderResult
{
    try {
        $cart = $this->cartPersister->load($cartToken, $context);
    } catch (CartTokenNotFoundException $e) {
        $cart = new Cart($context->getSalesChannel()->getTypeId(), $cartToken);
    }

    return $this->loadByCart($context, $cart, new CartBehavior($context->getPermissions()));
}
```

and instead it should do something like 

```
public function loadByToken(SalesChannelContext $context, string $cartToken): RuleLoaderResult
{
    try {
        $cart = $this->cartPersister->load($cartToken, $context);
    } catch (CartTokenNotFoundException $e) {
       $cart = $this->createNew($context,$token);
    }

    return $this->loadByCart($context, $cart, new CartBehavior($context->getPermissions()));
}
```

and new method added that evokes the event , then the event can be subscribed to and also method decorated if needed

**Comment by shyim** [comment]
Hey,

I don't see a benefit to create a new method for that. As this method is called inside the same class the decoration pattern will not fit there.

We would like to see a Pull Request from you when you want :)

**Comment by speedupmate** [comment]
Since its not a separate method you can't really know if you get a new cart or loaded cart from that method and theres no easy way to poke with it either. Rewriting a class and entire method should be a last resort (but then again in shopware it might be preferred)  

With decorator you can hack before/after calls if really needed since shopware lacks the before, after, around (I might be wrong again since I'm new to shopware) . From that point of view what fits or does not is for developers to decide, platform should provide the flexibility. 

I'm too new for shopware to make meaningful PR -s , lack the full overview and knowledge what is what and why. I only noticed this cause I needed to catch event when cart is created before it is saved to database first time , you easily can't do that in shopware (but I might be wrong). If it were up to me I would add it as a separate method as in:

`./Core/Checkout/Cart/SalesChannel/CartLoadRoute.php:88:        $cart = new Cart($name, $token);

What I also don't understand is if the cart is created in transport object it also loads all the rules, conditions etc even if it is a new cart object with no line-items added jet, waste of resources as with 0 cart there's nothing to calculate, discount (but I might be wrong, not seeing the full picture jet)

**Comment by OliverSkroblin** [comment]
> What I also don't understand is if the cart is created in transport object it also loads all the rules, conditions etc even if it is a new cart object with no line-items added jet, waste of resources as with 0 cart there's nothing to calculate, discount (but I might be wrong, not seeing the full picture jet)

We need the cart and all rules to validate which are matching, even for an empty cart. Rules may contain none related cart conditions.


> What I also don't understand is if the cart is created in transport object it also loads all the rules, conditions etc even if it is a new cart object with no line-items added jet, waste of resources as with 0 cart there's nothing to calculate, discount (but I might be wrong, not seeing the full picture jet)

You are right about that. However, there are places in Shopware where we don't want to allow decoration. 
This is mostly when we don't see a point in completely replacing the function or see too much of a risk to stability.
In this case we would be against the decoration pattern because otherwise a developer would have to customize several places to extend exactly the place when a new cart is created.

**Comment by speedupmate** [comment]
> In this case we would be against the decoration pattern because otherwise a developer would have to customise several places to extend exactly the place when a new cart is created. 

implement the missing event if possible as otherwise the only choice is a decorator on CartRuleLoader::loadByToken that detects the cart creation "event" if: 

* route/path = /checkout/line-item/add as only then the first object is saved by default
* rawtotal = 0
* lineitems = 0 

thats where your cart object is created and stored and the way to diff is to check if cart is created and not loaded is to check against the cart data. 

if the new cart creation would be inside a CartRuleLoader::loadByToken separate method `try/catch->createNew` instead of just` try/catch->new object`  then the check if its loaded vs created would be easier for others to understand even if decorated 

So to sum it up sw6 is missing a proper way to detect when cart is created and saved  for the first time. First save is important here cause cart is created {n} times in the transport object.

Performance wise I'm not sure (too new to sw) if simple object level parameter cache would save you from creating those objects over and over on that try/catch new construction since it all happens within same request context . Not sure if you even need a new object to be initiated all the times there as you do it on each request to server (+ ajax requests)

**Comment by JoshuaBehrens** [comment]
I see the technical entrypoints are missing, but what do you try to do with these 👀 ?

**Comment by speedupmate** [comment]
**For what that event would be useful for (https://en.wikipedia.org/wiki/Aspect-oriented_programming - before, after , around) ?**

* event on cart (or any object) that is created  and about to be stored in similar `try/load/catch/create` 

**what happens around those conditions:** 

* token is tied to guest cart first time
* cart defaults are stored first time 

**What you can do:**

* manipulate the cart object on creation with actions that need to be done only one time when real cart that is stored. This can be analysis or just creating separate objects whatever the feature at hand might need. 

`CartCreatedEvent` does not really fire in logical order and is either named oddly or just does not have a meaning as it does not cover all processes that are creating carts. I specifically need to know when the cart is created first time to launch actions only once. 

Cart object can change {n} times before its actually stored and after it is stored all those `try/load/catch/create ` will only load the right object (even if still {n} times). So theoretically in sw6 cart is not a cart before it is stored as it fires a exception (its not found in db) and right now there is no event `CartCreatedEvent ` that would fire when the cart is stored first time (eq when cart becomes real and does not throw exception any more for recreation) 
 
I agree thats not meaningful to evoke the event when this object is created as it is created loosely and not persisted always.

In my opinion if 

`\Shopware\Core\Checkout\Cart\CartRuleLoader::loadByToken ` 

gets

```
public function loadByToken(SalesChannelContext $context, string $cartToken): RuleLoaderResult
{
    try {
        $cart = $this->cartPersister->load($cartToken, $context);
    } catch (CartTokenNotFoundException $e) {
       $cart = $this->createNew($context,$token);
    }

    return $this->loadByCart($context, $cart, new CartBehavior($context->getPermissions()));
}
```

Solves the issue :
* event can be fired in separate method `createNew($context,$token)`
* same pattern is followed in other files where this event is fired
* separate method can be decorated to get before/after/around even if you don't prefer this and developers might need to do this 
* if anyone wants to go crazy and avoid decorating this then they can set up multiple event listeners and mark the events and try to follow the event triggering order to find out when the actual `CartCreatedEvent`  has followup events like items added etc

**Comment by MarcelSchmaeing** [comment]
Hello,

Within our company roadmap and work capacity, we try to address each bug or improvement request but admit that not each one will be resolved. To continue our culture of honesty and openness, we are closing this issue to focus on our roadmap on behalf of all Shopware users.

If you feel like this issue is still impacting you, please create a new issue and let us know.
Thank you again for your valuable feedback.
